File: 프로그래머스/2/81302. 거리두기 확인하기/거리두기 확인하기.cpp
#include <string>
#include <vector>
#include <queue>

using namespace std;

int xmoving[4] = {0, 1, 0, -1};
int ymoving[4] = {1, 0, -1, 0};

vector<int> solution(vector<vector<string>> places) {
    vector<int> answer;

    for (int seq = 0; seq < 5; ++seq) {  // 각 대기실(5개)에 대해 검사
        bool isSafe = true;

        for (int x = 0; x < 5 && isSafe; ++x) {
            for (int y = 0; y < 5 && isSafe; ++y) {
                if (places[seq][x][y] != 'P') continue;

                // BFS 시작
                int chk[5][5] = {0};
                queue<pair<int, int>> q;
                chk[x][y] = 1;
                q.push({x, y});

                while (!q.empty() && isSafe) {
                    auto [now_x, now_y] = q.front();
                    q.pop();

                    if (chk[now_x][now_y] >= 3) continue;

                    for (int i = 0; i < 4; i++) {
                        int tx = now_x + xmoving[i];
                        int ty = now_y + ymoving[i];

                        if (tx < 0 || ty < 0 || tx >= 5 || ty >= 5) continue;
                        if (chk[tx][ty] || places[seq][tx][ty] == 'X') continue;

                        if (places[seq][tx][ty] == 'P') {
                            isSafe = false; // 거리두기 위반
                            break;
                        }

                        chk[tx][ty] = chk[now_x][now_y] + 1;
                        q.push({tx, ty});
                    }
                }
            }
        }

        answer.push_back(isSafe ? 1 : 0);
    }

    return answer;
}

File: 프로그래머스/2/81302. 거리두기 확인하기/거리두기 확인하기.py
from collections import deque
from typing import List

def is_safe(place: List[str]) -> int:
    directions = [(1, 0), (-1, 0), (0, 1), (0, -1)]
    for r in range(5):
        for c in range(5):
            if place[r][c] != 'P':
                continue
            visited = [[False] * 5 for _ in range(5)]
            visited[r][c] = True
            queue = deque([(r, c, 0)])
            while queue:
                x, y, dist = queue.popleft()
                if dist >= 2:
                    continue
                for dx, dy in directions:
                    nx, ny = x + dx, y + dy
                    if not (0 <= nx < 5 and 0 <= ny < 5):
                        continue
                    if visited[nx][ny] or place[nx][ny] == 'X':
                        continue
                    if place[nx][ny] == 'P':
                        return 0
                    visited[nx][ny] = True
                    queue.append((nx, ny, dist + 1))
    return 1

def solution(places: List[List[str]]) -> List[int]:
    return [is_safe(place) for place in places]
